### PR TITLE
feat(behavior_velocity): update grid map collision judgement using object info

### PR DESCRIPTION
**Pull request by taikitanaka3**
## Description

This PR add following feature
- add processing time as debug option
- add occlusion spot radius(instead of line) to use collision free judgement considering space that pedestrian can hide 
- reflect dynamic object info to occupancy grid

This PR doesn't consider
- future movement of predicted object or lanelet that predicted object will occupy.
- roughness of the occupancy grid resolution.

## Related links

#258

## Tests performed

By Psim

## Notes for reviewers

small change
- use morphoEx instead of dilate erode and use cv values instead of occupancy grid value.
- early continue before using partition lanelet
- replace param and functions to  planning utils and planner data

## Pre-review checklist for the PR author

PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

Reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has the write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_velocity_planner/config/occlusion_spot.param.yaml [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
-      use_partition_lanelet: true          # [-] whether to use partition lanelet map data
-      pedestrian_vel: 1.0                  # [m/s] assume pedestrian is dashing from occlusion at this velocity
+      detection_method: "predicted_object"  # [-] candidate is "occupancy_grid" or "predicted_object"
+      pass_judge: "current_velocity"        # [-] candidate is "current_velocity"
+      filter_occupancy_grid: true           # [-] whether to filter occupancy grid by morphologyEx or not
+      use_object_info: true                 # [-] whether to reflect object info to occupancy grid map or not
+      use_partition_lanelet: true           # [-] whether to use partition lanelet map data
+      pedestrian_vel: 1.5                   # [m/s] assume pedestrian is dashing from occlusion at this velocity
+      pedestrian_radius: 0.5                # [m] assume pedestrian width(0.25m) + margin(0.25m)
+      debug:                                # !Note: default should be false for performance
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/config/occlusion_spot.param.yaml:11]: Note: separate debug parameters for utility

File: planning/behavior_velocity_planner/config/occlusion_spot.param.yaml [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+        is_show_occlusion: false            # [-] whether to show occlusion point markers.
+        is_show_cv_window: false            # [-] whether to show open_cv debug window
+        is_show_processing_time: false      # [-] whether to show processing time
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/config/occlusion_spot.param.yaml:14]: Note: separate debug parameters for utility

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/grid_utils.hpp [hunk 3]
```diff
@@ -34,12 +39,17 @@ namespace behavior_velocity_planner
...
-static constexpr int NO_INFORMATION = -1;
-static constexpr int FREE_SPACE = 0;
-static constexpr int UNKNOWN = 50;
-static constexpr int OCCUPIED = 100;
+static constexpr unsigned char FREE_SPACE = 0;
+static constexpr unsigned char UNKNOWN = 50;
+static constexpr unsigned char OCCUPIED = 100;
+static constexpr unsigned char FREE_SPACE_IMAGE = 0;
+static constexpr unsigned char UNKNOWN_IMAGE = 128;
+static constexpr unsigned char OCCUPIED_IMAGE = 255;
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/include/scene_module/occlusion_spot/grid_utils.hpp:52]: Note: use this value for better image processing

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/manager.cpp [hunk 4]
```diff
@@ -31,10 +31,6 @@ OcclusionSpotModuleManager::OcclusionSpotModuleManager(rclcpp::Node & node)
 : SceneModuleManagerInterface(node, getModuleName())
 {
   const std::string ns(getModuleName());
-  pub_debug_occupancy_grid_ =
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/manager.cpp:34]: Note: removed heavy debug publisher and use cv::imshow at debug

File: planning/behavior_velocity_planner/config/occlusion_spot.param.yaml [hunk 5]
```diff
@@ -1,11 +1,17 @@
 /**:
   ros__parameters:
     occlusion_spot:
-      detection_method: "predicted_object" # [-] candidate is "occupancy_grid" or "predicted_object"
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/config/occlusion_spot.param.yaml:4]: Note : change to occupancy_grid for testing

**Comment by taikitanaka3** [comment]
@yukkysaito @0x126 
I've added unit test to compare performance of polygon_vs_line_iterator
the result is as following and time may not differ much.
```
2: [ RUN      ] compareTime.polygon_vs_line_iterator
2: length of polygon: 0
2: polygon iterator [ms]: 0.116 :num: 0
2: Failed to construct LineIterator.
2: line iterator [ms]: 0.201 :num: 0
2: length of polygon: 1.41421
2: polygon iterator [ms]: 0.119 :num: 16
2: line iterator [ms]: 0.068 :num: 27
2: length of polygon: 2.82843
2: polygon iterator [ms]: 0.124 :num: 31
2: line iterator [ms]: 0.053 :num: 54
2: length of polygon: 4.24264
2: polygon iterator [ms]: 0.108 :num: 45
2: line iterator [ms]: 0.036 :num: 81
2: length of polygon: 5.65685
2: polygon iterator [ms]: 0.156 :num: 60
2: line iterator [ms]: 0.081 :num: 108
2: length of polygon: 7.07107
2: polygon iterator [ms]: 0.137 :num: 73
2: line iterator [ms]: 0.085 :num: 135
2: length of polygon: 8.48528
2: polygon iterator [ms]: 0.138 :num: 86
2: line iterator [ms]: 0.041 :num: 153
2: length of polygon: 9.89949
2: polygon iterator [ms]: 0.192 :num: 101
2: line iterator [ms]: 0.042 :num: 180
2: length of polygon: 11.3137
2: polygon iterator [ms]: 0.2 :num: 116
2: line iterator [ms]: 0.042 :num: 207
2: length of polygon: 12.7279
2: polygon iterator [ms]: 0.234 :num: 130
2: line iterator [ms]: 0.103 :num: 234
```

File: planning/behavior_velocity_planner/test/src/test_grid_utils.cpp [hunk 6]
```diff
@@ -101,3 +104,113 @@ TEST(isOcclusionSpotSquare, occlusion_spot_cell)
...
+            count++;
+            // continue
+          }
+        }
+      } catch (const std::invalid_argument & e) {
+        std::cerr << e.what() << std::endl;
+      }
+    }
+    std::cout << "polygon iterator [ms]: " << stop_watch.toc("processing_time", true)
+              << " :num: " << count << std::endl;
```
Comment (yukkysaito) [inline planning/behavior_velocity_planner/test/src/test_grid_utils.cpp:184]: pls remove

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/manager.cpp [hunk 7]
```diff
@@ -62,19 +58,26 @@ OcclusionSpotModuleManager::OcclusionSpotModuleManager(rclcpp::Node & node)
...
+  pp.use_object_info = node.declare_parameter(ns + ".use_object_info", false);
   pp.use_partition_lanelet = node.declare_parameter(ns + ".use_partition_lanelet", false);
-  pp.pedestrian_vel = node.declare_parameter(ns + ".pedestrian_vel", 1.0);
+  pp.pedestrian_vel = node.declare_parameter(ns + ".pedestrian_vel", 1.5);
+  pp.pedestrian_radius = node.declare_parameter(ns + ".pedestrian_radius", 0.5);
+
+  // debug
+  pp.is_show_occlusion = node.declare_parameter(ns + ".debug.is_show_occlusion", false);
+  pp.is_show_cv_window = node.declare_parameter(ns + ".debug.is_show_cv_window", false);
+  pp.is_show_processing_time = node.declare_parameter(ns + ".debug.is_show_processing_time", false);
```
Comment (yukkysaito) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/manager.cpp:70]: I feel commas should not be used.

Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/manager.cpp:70]: @yukkysaito 
can you explain more clearly?
Do you want to remove default arguments or something else?

File: planning/behavior_velocity_planner/test/src/test_grid_utils.cpp (hunk 6 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/test/src/test_grid_utils.cpp]: @yukkysaito 
I make it silent at [761e310](https://github.com/autowarefoundation/autoware.universe/pull/568/commits/761e31039efd8fe26e133e9317c1fa9e3537fb8b)

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/manager.cpp (hunk 7 above)
Comment (yukkysaito) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/manager.cpp:70]: sorry, i misstaked

**Comment by taikitanaka3** [comment]
@0x126 
[7c2ad42](https://github.com/autowarefoundation/autoware.universe/pull/568/commits/7c2ad42738700fede3345bd676fda6c856422f6c)
is going to look like
image

**Review by 0x126 — APPROVED** [review]
LGTM

**Pull request merged**